### ISSUE TITLE
clippy::borrowed_box shouldn't want on return type

### ISSUE THREAD

**Issue by KSXGitHub** [rust-lang/rust-clippy#10982]
### Summary

This function should be fine as the return type not only derefs to `&[Item]` but also allow more operations than just `&[Item]`:

```rust
fn inner(&self) -> &'_ Box<[Item]> {
    &self.0
}
```

### Lint Name

borrowed_box

### Reproducer

I tried this code:

```rust
fn inner(&self) -> &'_ Box<[Item]> {
    &self.0
}
```

I saw this happen:

```
you seem to be trying to use `&Box<T>`. Consider using just `&T`
for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#borrowed_box
`#[warn(clippy::borrowed_box)]` on by default
```

I expected to see this happen:

No warnings

### Version

```text
rustc 1.70.0 (90c541806 2023-05-31)
binary: rustc
commit-hash: 90c541806f23a127002de5b4038be731ba1458ca
commit-date: 2023-05-31
host: x86_64-unknown-linux-gnu
release: 1.70.0
LLVM version: 16.0.2
```

**Comment by y21** [comment]
What can you do with a `&Box<[T]>` that you couldn't do with a `&[T]`?

**Comment by KSXGitHub** [comment]
> What can you do with a `&Box<[T]>` that you couldn't do with a `&[T]`?

Passing it to a function that expects a reference to a `Box` for example. You'll never know with generics.

**Comment by blyxyas** [comment]
Wait, @y21 are you currently working on this? Should I start working or will you submit a PR?

**Comment by y21** [comment]
nope, I'm not working on it. You can take it if you want 😄

**Comment by Nilstrieb** [comment]
I'm not sure whether that is enough justification for disabling the lint on return types. The same argument can be made for argument types as well. Maybe the body needs to pass it to another function that expects it. I think it's okay for the lint to have "false positives" in these cases where unidiomatic types occur (unless clippy can know directly that this is because of an external requirement, which it cannot for function return types).

**Comment by blyxyas** [comment]
I'm also not sure how to handle this. I'll leave the PR open, waiting for some conclusion about it.
